### DISCUSSION TITLE
Open-attestation on the Polygon network

### DISCUSSION THREAD

**Discussion by gsyndromes** [Open-Attestation/open-attestation#182] [Ideas]
Hi! In light with the many blockchains popping off and their reliability ever-increasing with increasing user bases, it would be super helpful that open-attestation moves along with the innovations and allow its functions to operate on other blockchains as well. I hope to be able to see all functions of Open-attestation being supported on the polygon network due to the **significantly** lower gas fees available on polygon. It makes it easier for projects to onboard open-attestation as well if the functions are is blockchain-agnostic.

**Comment by Nebulis** [comment]
As I mentioned, it will be nice if you could give us more information about how open-attestation could support polygon.

You are not the first one to request to support a different blockchain. But definitely, if we have to explore every options, we won't go anywhere.

**Reply by gsyndromes** [reply]
Ah, it would be great if we could use the open-attestation-cli to run on the polygon chain, for example, in issuing the document-store onto the blockchain, the current choices for the --network field are only mainnet, ropsten, rinkeby and local as seen in the photo below.
![image](https://user-images.githubusercontent.com/44989392/122147998-76defa80-ce8c-11eb-890d-3c6161694426.png)

**Reply by gsyndromes** [reply]
And also, being able to verify these documents on tradetrust/opencerts would be a great help too.

**Reply by Nebulis** [reply]
You explained the features you want us to provide. You don't explain all the work that needs to be done or how we could reuse the existing work and adapt it to use polygon chain.

A simple example:
We have a [smart contract](https://github.com/Open-Attestation/document-store) for ethereum. Can we reuse it for polygon? If can't then do we need to develop another smart contract?

Integrating with another blockchain is not as easy as just saying it.

**Reply by gsyndromes** [reply]
According to this article by MATIC, 
https://docs.matic.network/docs/integrate/quickstart/
it is possible by changing the RPC URL to their testnet and deploying the contract onto the polygon network using remix or truffle. Is this what you mean?

**Reply by Nebulis** [reply]
If that's the case then can you try directly with [oa-verify](https://github.com/Open-Attestation/oa-verify).

When you build the verifier you can provide a `provider` property, which allows you to specify any RPC URL.

IF that works, that means there is very little to do to support polygon.

**Reply by gsyndromes** [reply]
hi how do i test this command?

import { documentRopstenValidWithToken } from "./test/fixtures/v2/documentRopstenValidWithToken";
import { verify, isValid } from "@govtechsg/oa-verify";

const fragments = await verify(documentRopstenValidWithToken);
console.log(fragments); // see below
console.log(isValid(fragments)); // display true

**Reply by Nebulis** [reply]
Run it into node is or ts node

You need a custom provider like https://www.github.com/Open-Attestation/verify.gov.sg/tree/19a5fd61fcc92d3628a9e16e95864f0e1ba68a3f/src%2Fissuers-verifier.ts (at the bottom of the file)

**Reply by gsyndromes** [reply]
so this is my entire file and I run it into ts-node?
``` 
import { documentRopstenValidWithToken } from "./test/fixtures/v2/documentRopstenValidWithToken";
import { verify, isValid } from "@govtechsg/oa-verify";

const fragments = await verify(documentRopstenValidWithToken);
console.log(fragments); // see below
console.log(isValid(fragments)); // display true



import {  CodedError,  createResolver,  ErrorVerificationFragment,  InvalidVerificationFragment,  openAttestationVerifiers,  SkippedVerificationFragment,  ValidVerificationFragment,  verificationBuilder,  VerificationFragmentType,  Verifier,} from "@govtechsg/oa-verify";import { getData, utils } from "@govtechsg/open-attestation";import { providers } from "ethers";
type AllowedIssuersValidFragment = ValidVerificationFragment&lt;Array&lt;string | undefined&gt;&gt;;type AllowedIssuersInvalidFragment = InvalidVerificationFragment&lt;Array&lt;string | undefined&gt;&gt;;type AllowedIssuersErrorFragment = ErrorVerificationFragment&lt;any&gt;;
type AllowedIssuersFragment =  | AllowedIssuersValidFragment  | AllowedIssuersErrorFragment  | AllowedIssuersInvalidFragment  | SkippedVerificationFragment;
type VerifierType = Verifier&lt;AllowedIssuersFragment&gt;;
export enum VerifyAllowedIssuersCode {  UNEXPECTED_ERROR = 0,  INVALID_IDENTITY = 1,  SKIPPED = 2,  UNSUPPORTED_V3_DOCUMENT = 3,}
const name = "VerifyAllowedIssuers";const type: VerificationFragmentType = "ISSUER_IDENTITY";const whitelistedIssuers = process.env.REACT_APP_WHITELISTED_ISSUERS?.split(",") || ["gov.sg", "openattestation.com"];export const isWhitelisted = (identity: string): boolean =&gt; {  return (    whitelistedIssuers.some((issuer) =&gt; identity.toLowerCase().endsWith(`.${issuer}`)) ||    whitelistedIssuers.includes(identity.toLowerCase())  );};
const skip: VerifierType["skip"] = () =&gt; {  throw new Error("This verifier is never skipped");};
const test: VerifierType["test"] = () =&gt; {  return true;};
const verifyMethod: VerifierType["verify"] = async (document) =&gt; {  try {    if (utils.isWrappedV2Document(document)) {      const documentData = getData(document);      const identities = documentData.issuers.map((issuer) =&gt; issuer.identityProof?.location);      // every issuers must be whitelisted      const valid =        identities.length &gt; 0 &amp;&amp; identities.every((identity) =&gt; (identity ? isWhitelisted(identity) : false));      if (!valid) {        return {          name,          type,          data: identities,          status: "INVALID" as const,          reason: {            code: VerifyAllowedIssuersCode.INVALID_IDENTITY,            codeString: VerifyAllowedIssuersCode[VerifyAllowedIssuersCode.INVALID_IDENTITY],            message: `No issuers allowed by this platform found. Valid issuers are ${whitelistedIssuers.join(",")}`,          },        };      }      return {        name,        type,        data: identities,        status: "VALID" as const,      };    }    // TODO: support for V3 is available now    throw new CodedError(      "Verify does not support v3 document",      VerifyAllowedIssuersCode.UNSUPPORTED_V3_DOCUMENT,      VerifyAllowedIssuersCode[VerifyAllowedIssuersCode.UNSUPPORTED_V3_DOCUMENT]    );  } catch (e) {    return {      name,      type,      data: e,      reason: {        code: e.code || VerifyAllowedIssuersCode.UNEXPECTED_ERROR,        codeString: e.codeString || VerifyAllowedIssuersCode[VerifyAllowedIssuersCode.UNEXPECTED_ERROR],        message: e.message,      },      status: "ERROR" as const,    };  }};
export const verifyAllowedIssuers: VerifierType = {  skip,  test,  verify: verifyMethod,};const NETWORK_NAME = process.env.REACT_APP_NETWORK_NAME || "ropsten";const INFURA_API_KEY = process.env.REACT_APP_INFURA_API_KEY;export const verify = verificationBuilder([...openAttestationVerifiers, verifyAllowedIssuers], {  provider: new providers.InfuraProvider(NETWORK_NAME, INFURA_API_KEY),  resolver: INFURA_API_KEY    ? createResolver({        ethrResolverConfig: {          networks: [{ name: "mainnet", rpcUrl: `https://mainnet.infura.io/v3/${INFURA_API_KEY}` }],        },      })    : undefined,});
```


i am receiving some errors doing so, using the command ```ts-node file_name.ts```

**Reply by Nebulis** [reply]
You are doing it wrong. You don't need to copy the full file content. You just need the provider option:

```ts
const verify = verificationBuilder([...openAttestationVerifiers, verifyAllowedIssuers], {
  provider: new providers.InfuraProvider(NETWORK_NAME, INFURA_API_KEY)
});
```

Instead of InfuraProvider, you might want to use a [custom json rpc provider](https://docs.ethers.io/v5/api/providers/other/#UrlJsonRpcProvider).

If you are not familiar with node.js or typescript it might be hard for you to try out. Maybe you can instead clone the repository, create a test and run it. The codebase is still in typescript though

**Reply by gsyndromes** [reply]
hmm okay i think i will need some time to try it out. thank you for your guidance. may i ask since this is the verification and before the document is verifiable it has to be issued onto the blockchain first. could i have some guidance on how should i go about doing that?

**Reply by ryanseee** [reply]
hi @garysyndromes,

you could follow these steps to try what you were asking for;

1) create an empty typescript project (you may need to figure out how to run a typescript file, the command would be, for e.g '_ts-node test.ts_')
2) install _@govtechsg/oa-verify_ package
3) install _ethers_ package
4) import _verificationBuilder_ and _openAttestationVerifiers_ from @govtechsg/oa-verify
5) import _providers_ from ethers
6) create a custom verification method using _oa-verify's_ verificationBuilder component like what @Nebulis mentioned earlier (using the component you imported in step 4). For the provider field that you have to supply, you can try _providers.JsonRpcProviders(<enter rpc url in string>)_ (using component you imported in step 5). 
7) if up to (6) doesnt work, it could be because you need to include 'Polygon PoS add on' in your infura account to be able to access Polygon's mainnet or testnet.
8) deploy the [DocumentStore.sol](https://github.com/Open-Attestation/document-store/blob/master/contracts/DocumentStore.sol) smart contract in [remix](https://remix.ethereum.org/).
9) _Issue_ a document(byte32). You should do this only after wrapping a raw document (maybe just try wrapping an old verifiable document but remember to change the document store address to what you deployed in step (8). After you get the hash of the document _issue_ it on remix.
10) use the raw document from step (9) and throw it in to the custom verifier from step (6) for verification.

P.S. If you face any issue along the way (which i believe would be the RPC part), you could probably try establishing the connection to your RPC url in a separate empty project before proceeding further.

**Reply by gsyndromes** [reply]
thanks @ryanseee for the detailed step by step guide! I seem to be having troubles with the compatibility issues with typescript and node regarding the import commands - wondering if there's a standard config to follow to be able to run OA scripts?

**Reply by ryanseee** [reply]
@garysyndromes what compatibility issues are you facing? to import the components from _@govtechsg/oa-verify_, you would need _typescript_ at the very least. importing these packages, for e.g. step (4), in your empty project should be quite straight forward.

**Reply by gsyndromes** [reply]
i think there is some config thats needed to be done to package.json and tsconfig.json to be able to import from directory 
`import { verificationBuilder, openAttestationVerifiers } from "@govtechsg/oa-verify/";`
properly. 
getting this error right now :
`  const err = new ERR_UNSUPPORTED_DIR_IMPORT(path, fileURLToPath(base)); `

still haven't been able to figure it out though

**Reply by ryanseee** [reply]
i see there is an extra forward slash in your _@govtechsg/oa-verify_, could that be the reason? otherwise, you could try replacing your entire content in tsconfig.json with the content below.

```
{
  "compilerOptions": {
    /* Visit https://aka.ms/tsconfig.json to read more about this file */

    /* Basic Options */
    // "incremental": true,                         /* Enable incremental compilation */
    "target": "es5" /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', 'ES2021', or 'ESNEXT'. */,
    "module": "commonjs" /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */,
    // "lib": [],                                   /* Specify library files to be included in the compilation. */
    // "allowJs": true,                             /* Allow javascript files to be compiled. */
    // "checkJs": true,                             /* Report errors in .js files. */
    // "jsx": "preserve",                           /* Specify JSX code generation: 'preserve', 'react-native', 'react', 'react-jsx' or 'react-jsxdev'. */
    // "declaration": true,                         /* Generates corresponding '.d.ts' file. */
    // "declarationMap": true,                      /* Generates a sourcemap for each corresponding '.d.ts' file. */
    // "sourceMap": true,                           /* Generates corresponding '.map' file. */
    // "outFile": "./",                             /* Concatenate and emit output to single file. */
    // "outDir": "./",                              /* Redirect output structure to the directory. */
    // "rootDir": "./",                             /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
    // "composite": true,                           /* Enable project compilation */
    // "tsBuildInfoFile": "./",                     /* Specify file to store incremental compilation information */
    // "removeComments": true,                      /* Do not emit comments to output. */
    // "noEmit": true,                              /* Do not emit outputs. */
    // "importHelpers": true,                       /* Import emit helpers from 'tslib'. */
    // "downlevelIteration": true,                  /* Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'. */
    // "isolatedModules": true,                     /* Transpile each file as a separate module (similar to 'ts.transpileModule'). */

    /* Strict Type-Checking Options */
    "strict": true /* Enable all strict type-checking options. */,
    // "noImplicitAny": true,                       /* Raise error on expressions and declarations with an implied 'any' type. */
    // "strictNullChecks": true,                    /* Enable strict null checks. */
    // "strictFunctionTypes": true,                 /* Enable strict checking of function types. */
    // "strictBindCallApply": true,                 /* Enable strict 'bind', 'call', and 'apply' methods on functions. */
    // "strictPropertyInitialization": true,        /* Enable strict checking of property initialization in classes. */
    // "noImplicitThis": true,                      /* Raise error on 'this' expressions with an implied 'any' type. */
    // "alwaysStrict": true,                        /* Parse in strict mode and emit "use strict" for each source file. */

    /* Additional Checks */
    // "noUnusedLocals": true,                      /* Report errors on unused locals. */
    // "noUnusedParameters": true,                  /* Report errors on unused parameters. */
    // "noImplicitReturns": true,                   /* Report error when not all code paths in function return a value. */
    // "noFallthroughCasesInSwitch": true,          /* Report errors for fallthrough cases in switch statement. */
    // "noUncheckedIndexedAccess": true,            /* Include 'undefined' in index signature results */
    // "noImplicitOverride": true,                  /* Ensure overriding members in derived classes are marked with an 'override' modifier. */
    // "noPropertyAccessFromIndexSignature": true,  /* Require undeclared properties from index signatures to use element accesses. */

    /* Module Resolution Options */
    // "moduleResolution": "node",                  /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
    // "baseUrl": "./",                             /* Base directory to resolve non-absolute module names. */
    // "paths": {},                                 /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
    // "rootDirs": [],                              /* List of root folders whose combined content represents the structure of the project at runtime. */
    // "typeRoots": [],                             /* List of folders to include type definitions from. */
    // "types": [],                                 /* Type declaration files to be included in compilation. */
    // "allowSyntheticDefaultImports": true,        /* Allow default imports from modules with no default export. This does not affect code emit, just typechecking. */
    "esModuleInterop": true /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */,
    // "preserveSymlinks": true,                    /* Do not resolve the real path of symlinks. */
    // "allowUmdGlobalAccess": true,                /* Allow accessing UMD globals from modules. */

    /* Source Map Options */
    // "sourceRoot": "",                            /* Specify the location where debugger should locate TypeScript files instead of source locations. */
    // "mapRoot": "",                               /* Specify the location where debugger should locate map files instead of generated locations. */
    // "inlineSourceMap": true,                     /* Emit a single file with source maps instead of having a separate file. */
    // "inlineSources": true,                       /* Emit the source alongside the sourcemaps within a single file; requires '--inlineSourceMap' or '--sourceMap' to be set. */

    /* Experimental Options */
    // "experimentalDecorators": true,              /* Enables experimental support for ES7 decorators. */
    // "emitDecoratorMetadata": true,               /* Enables experimental support for emitting type metadata for decorators. */

    /* Advanced Options */
    "skipLibCheck": true /* Skip type checking of declaration files. */,
    "forceConsistentCasingInFileNames": true /* Disallow inconsistently-cased references to the same file. */,
    "outDir": "dist",
    "sourceMap": true,
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "esModuleInterop": true
  }
}
```

**Reply by gsyndromes** [reply]
thank you for the example! It helped solved my import problem. by the way, apparently, typescript does not play well with the top-level await as per the readme. It throws this error  
` test.ts:20:20 - error TS1378: Top-level 'await' expressions are only allowed when the 'module' option is set to 'esnext' or 'system', and the 'target' option is set to 'es2017' or higher. `

` 20  const fragments = await verify(documentRopstenValidWithCertificateStore); `


which can be hacked through by replacing with 
` const fragments = async () => {
     await verify(documentRopstenValidWithCertificateStore);
 }; `



EDIT: the typescript version I am on is 
"typescript": "^4.3.4"

**Reply by gsyndromes** [reply]
however, upon running it, i am seeing this error instead, do you have any ideas why?
`Object.defineProperty(exports, "__esModule", { value: true });
                       ^
 ReferenceError: exports is not defined `

**Reply by Nebulis** [reply]
@garysyndromes Ryan, who helped you, manage to answer my question: we are able to use oa-verify directly with polygon without any changes.

That means that oa supports natively polygon. However, OpenCerts and Tradetrust do not (we need special providers). It's a matter of configuration.
If we want Opencerts and Tradetrust to support both Eth / Polygon, we will need to work on OA to allow them to decide where to perform a verification.

We don't think that Polygon is mature enough. We don't plan to do modifications now. However, we will keep an eye on it and if the time comes we will do it.

There are things that can be done for now:
- Nothing prevents you to create your own platform that supports Polygon or Eth (Or both, considering Infura supports both)
- We can already start to think about how to integrate with different providers. Feel free to join the conversation

**Reply by gsyndromes** [reply]
Can i clarify what do you mean by opencerts and tradetrust needing special providers?

**Reply by ryanseee** [reply]
@garysyndromes what laurent meant was that, if you take a look at https://www.tradetrust.io/, https://dev.www.tradetrust.io/ and https://rinkeby.www.tradetrust.io/, they are all running on different Ethereum networks. so to include Polygon Mainnet or Mumbai Testnet necessary configurations would need to be done to support them on the different projects/platforms.
